File: frontend/lib/models/autonomy.dart
import 'package:flutter/foundation.dart';

String? _string(Object? value) => value is String ? value : value?.toString();
DateTime? _date(Object? value) {
  final text = _string(value);
  if (text == null) return null;
  return DateTime.tryParse(text)?.toUtc();
}

Map<String, dynamic> _map(Object? value) =>
    value is Map ? Map<String, dynamic>.from(value) : <String, dynamic>{};
List<String> _strings(Object? value) => value is List
    ? value.map(_string).whereType<String>().toList()
    : <String>[];
String _enum(Object? value, String fallback) => _string(value) ?? fallback;

@immutable
class Agent {
  final String id, name, status, executionMode, threadId;
  final String handle;
  final String? threadTitle;
  final bool isModerator;
  final String? description, activeVersionId, templateId;
  final int concurrencyLimit, queueLimit;
  final DateTime? createdAt, updatedAt;
  const Agent({
    required this.id,
    required this.name,
    required this.status,
    required this.executionMode,
    required this.threadId,
    this.handle = '',
    this.threadTitle,
    this.isModerator = false,
    this.description,
    this.activeVersionId,
    this.templateId,
    this.concurrencyLimit = 1,
    this.queueLimit = 100,
    this.createdAt,
    this.updatedAt,
  });
  factory Agent.fromJson(Map<String, dynamic> j) => Agent(
    id: _string(j['id']) ?? '',
    name: _string(j['name']) ?? 'Unnamed agent',
    status: _enum(j['status'], 'unknown'),
    executionMode: _enum(j['execution_mode'], 'unknown'),
    threadId: _string(j['thread_id']) ?? '',
    handle: _string(j['handle'] ?? j['mention_name']) ?? '',
    threadTitle: _string(j['thread_title']),
    isModerator: j['is_moderator'] == true,
    description: _string(j['description']),
    activeVersionId: _string(j['active_version_id']),
    templateId: _string(j['template_id']),
    concurrencyLimit: j['concurrency_limit'] is num
        ? (j['concurrency_limit'] as num).toInt()
        : 1,
    queueLimit: j['queue_limit'] is num
        ? (j['queue_limit'] as num).toInt()
        : 100,
    createdAt: _date(j['created_at']),
    updatedAt: _date(j['updated_at']),
  );
}

class Draft {
  final String id, agentId, configHash, promptTemplate;
  final int optimisticLockVersion, schemaVersion;
  final Map<String, dynamic> config;
  final List<String> toolSelection, skillSelection;
  final List<Map<String, dynamic>> credentialBindings;
  final DateTime? updatedAt;
  const Draft({
    this.id = '',
    this.agentId = '',
    this.configHash = '',
    this.promptTemplate = '',
    this.optimisticLockVersion = 1,
    this.schemaVersion = 1,
    this.config = const {},
    this.toolSelection = const [],
    this.skillSelection = const [],
    this.credentialBindings = const [],
    this.updatedAt,
  });
  factory Draft.fromJson(Map<String, dynamic> j) => Draft(
    id: _string(j['id']) ?? '',
    agentId: _string(j['agent_id']) ?? '',
    configHash: _string(j['config_hash']) ?? '',
    promptTemplate: _string(j['prompt_template']) ?? '',
    optimisticLockVersion: (j['optimistic_lock_version'] ?? j['version']) is num
        ? ((j['optimistic_lock_version'] ?? j['version']) as num).toInt()
        : 1,
    schemaVersion: j['schema_version'] is num
        ? (j['schema_version'] as num).toInt()
        : 1,
    config: _map(j['config']),
    toolSelection: _strings(j['tool_selection']),
    skillSelection: _strings(j['skill_selection']),
    credentialBindings: (j['credential_bindings'] is List)
        ? (j['credential_bindings'] as List)
              .whereType<Map>()
              .map((x) => Map<String, dynamic>.from(x))
              .toList()
        : const [],
    updatedAt: _date(j['updated_at']),
  );
}

class Version {
  final String id, agentId, configHash, promptTemplate;
  final int version;
  final Map<String, dynamic> config;
  final List<String> toolSelection, skillSelection;
  const Version({
    this.id = '',
    this.agentId = '',
    this.configHash = '',
    this.promptTemplate = '',
    this.version = 0,
    this.config = const {},
    this.toolSelection = const [],
    this.skillSelection = const [],
  });
  factory Version.fromJson(Map<String, dynamic> j) => Version(
    id: _string(j['id']) ?? '',
    agentId: _string(j['agent_id']) ?? '',
    configHash: _string(j['config_hash']) ?? '',
    promptTemplate: _string(j['prompt_template']) ?? '',
    version: j['version'] is num ? (j['version'] as num).toInt() : 0,
    config: _map(j['config']),
    toolSelection: _strings(j['tool_selection']),
    skillSelection: _strings(j['skill_selection']),
  );
}

class Trigger {
  final String id, type;
  final Map<String, dynamic> config;
  final bool active;
  const Trigger({
    this.id = '',
    this.type = 'unknown',
    this.config = const {},
    this.active = true,
  });
  factory Trigger.fromJson(Map<String, dynamic> j) => Trigger(
    id: _string(j['id']) ?? '',
    type: _enum(j['trigger_type'], 'unknown'),
    config: _map(j['config']),
    active: j['is_active'] != false,
  );
}

class Run {
  final String id, agentId, agentVersionId, threadId, status, mode;
  final String route;
  final String? inputMessageId;
  final String? agentName, agentHandle;
  final String? outputSummary, triggerEventId;
  final DateTime? queuedAt, startedAt, completedAt;
  const Run({
    this.id = '',
    this.agentId = '',
    this.agentVersionId = '',
    this.threadId = '',
    this.status = 'unknown',
    this.mode = 'unknown',
    this.route = '',
    this.inputMessageId,
    this.agentName,
    this.agentHandle,
    this.outputSummary,
    this.triggerEventId,
    this.queuedAt,
    this.startedAt,
    this.completedAt,
  });
  factory Run.fromJson(Map<String, dynamic> j) => Run(
    id: _string(j['id']) ?? '',
    agentId: _string(j['agent_id']) ?? '',
    agentVersionId: _string(j['agent_version_id']) ?? '',
    threadId: _string(j['thread_id']) ?? '',
    status: _enum(j['status'], 'unknown'),
    mode: _enum(j['mode'], 'unknown'),
    route: _string(j['route']) ?? '',
    inputMessageId: _string(j['input_message_id'] ?? j['inputMessageId']),
    agentName: _string(j['agent_name']),
    agentHandle: _string(j['agent_handle'] ?? j['handle']),
    outputSummary: _string(j['output_summary']),
    triggerEventId: _string(j['trigger_event_id']),
    queuedAt: _date(j['queued_at']),
    startedAt: _date(j['started_at']),
    completedAt: _date(j['completed_at']),
  );

  Run copyWith({
    String? inputMessageId,
    String? agentName,
    String? agentHandle,
  }) => Run(
    id: id,
    agentId: agentId,
    agentVersionId: agentVersionId,
    threadId: threadId,
    status: status,
    mode: mode,
    route: route,
    inputMessageId: inputMessageId ?? this.inputMessageId,
    agentName: agentName ?? this.agentName,
    agentHandle: agentHandle ?? this.agentHandle,
    outputSummary: outputSummary,
    triggerEventId: triggerEventId,
    queuedAt: queuedAt,
    startedAt: startedAt,
    completedAt: completedAt,
  );
}

@immutable
class ActiveRunPresentation {
  final Run run;
  final List<RunEvent> events;

  const ActiveRunPresentation({required this.run, this.events = const []});

  String get id => run.id;
  String? get inputMessageId => run.inputMessageId;
}

class RunEvent {
  final int sequence;
  final String type;
  final Map<String, dynamic> payload;
  final DateTime? createdAt;
  const RunEvent({
    this.sequence = 0,
    this.type = 'unknown',
    this.payload = const {},
    this.createdAt,
  });
  factory RunEvent.fromJson(Map<String, dynamic> j) => RunEvent(
    sequence: j['sequence'] is num
        ? (j['sequence'] as num).toInt()
        : int.tryParse(_string(j['cursor']) ?? '0') ?? 0,
    type: _enum(j['event_type'] ?? j['type'], 'unknown'),
    payload: _map(j['payload']),
    createdAt: _date(j['created_at']),
  );
}

class AuditEntry {
  final String id, type, resourceType;
  final Map<String, dynamic> metadata;
  final DateTime? createdAt;
  const AuditEntry({
    this.id = '',
    this.type = 'unknown',
    this.resourceType = 'unknown',
    this.metadata = const {},
    this.createdAt,
  });
  factory AuditEntry.fromJson(Map<String, dynamic> j) => AuditEntry(
    id: _string(j['id']) ?? '',
    type: _enum(j['event_type'], 'unknown'),
    resourceType: _enum(j['resource_type'], 'unknown'),
    metadata: _map(j['metadata'] ?? j['payload']),
    createdAt: _date(j['created_at']),
  );
}

@immutable
class HeartbeatStatus {
  final String agentId;
  final bool enabled;
  final int minWakeSeconds;
  final int maxWakeSeconds;
  final double idleBackoffFactor;
  final int revision;
  final String operationalStatus;
  final String? workflowId;
  final DateTime? lastWakeAt;
  final DateTime? lastCompletedAt;
  final DateTime? nextWakeAt;
  final String? lastDecision;
  final String? lastRunId;
  final int consecutiveNoops;
  final String? lastError;
  final DateTime? updatedAt;
  const HeartbeatStatus({
    required this.agentId,
    this.enabled = false,
    this.minWakeSeconds = 300,
    this.maxWakeSeconds = 3600,
    this.idleBackoffFactor = 2.0,
    this.revision = 1,
    this.operationalStatus = 'disabled',
    this.workflowId,
    this.lastWakeAt,
    this.lastCompletedAt,
    this.nextWakeAt,
    this.lastDecision,
    this.lastRunId,
    this.consecutiveNoops = 0,
    this.lastError,
    this.updatedAt,
  });
  factory HeartbeatStatus.fromJson(Map<String, dynamic> j) {
    final backoff = j['idle_backoff_factor'];
    return HeartbeatStatus(
      agentId: _string(j['agent_id']) ?? '',
      enabled: j['enabled'] is bool ? j['enabled'] as bool : false,
      minWakeSeconds: j['min_wake_seconds'] is num
          ? (j['min_wake_seconds'] as num).toInt()
          : 300,
      maxWakeSeconds: j['max_wake_seconds'] is num
          ? (j['max_wake_seconds'] as num).toInt()
          : 3600,
      idleBackoffFactor: backoff is num
          ? (backoff).toDouble()
          : double.tryParse(_string(backoff) ?? '2.0') ?? 2.0,
      revision: j['revision'] is num ? (j['revision'] as num).toInt() : 1,
      operationalStatus: _enum(j['operational_status'], 'disabled'),
      workflowId: _string(j['workflow_id']),
      lastWakeAt: _date(j['last_wake_at']),
      lastCompletedAt: _date(j['last_completed_at']),
      nextWakeAt: _date(j['next_wake_at']),
      lastDecision: _string(j['last_decision']),
      lastRunId: _string(j['last_run_id']),
      consecutiveNoops: j['consecutive_noops'] is num
          ? (j['consecutive_noops'] as num).toInt()
          : 0,
      lastError: _string(j['last_error']),
      updatedAt: _date(j['updated_at']),
    );
  }
  String get statusLabel {
    switch (operationalStatus) {
      case 'disabled':
        return 'Off';
      case 'scheduled':
        return nextWakeAt != null
            ? 'Next wake ${_relative(nextWakeAt!)}'
            : 'Scheduled';
      case 'evaluating':
        return 'Evaluating';
      case 'paused':
        return 'Paused';
      case 'blocked_mode':
        return 'Blocked (thread not agent mode)';
      case 'blocked_archived':
        return 'Blocked (archived)';
      case 'blocked_global':
        return 'Blocked (autonomy disabled)';
      case 'error':
        return 'Error';
      default:
        return operationalStatus;
    }
  }

  String _relative(DateTime target) {
    final delta = target.difference(DateTime.now().toUtc());
    if (delta.isNegative) return 'now';
    final mins = delta.inMinutes;
    if (mins < 1) return 'in ${delta.inSeconds}s';
    if (mins < 60) return 'in $mins min';
    final hours = delta.inHours;
    if (hours < 24) return 'in ${hours}h';
    return 'in ${delta.inDays}d';
  }
}

class HeartbeatConfig {
  final bool enabled;
  final int minWakeSeconds;
  final int maxWakeSeconds;
  final double idleBackoffFactor;
  final int? expectedRevision;
  const HeartbeatConfig({
    required this.enabled,
    this.minWakeSeconds = 300,
    this.maxWakeSeconds = 3600,
    this.idleBackoffFactor = 2.0,
    this.expectedRevision,
  });
  Map<String, dynamic> toJson() => {
    'enabled': enabled,
    'min_wake_seconds': minWakeSeconds,
    'max_wake_seconds': maxWakeSeconds,
    'idle_backoff_factor': idleBackoffFactor,
    if (expectedRevision != null) 'expected_revision': expectedRevision,
  };
}

File: backend/app/contracts/autonomy.py
"""Public Phase 1 autonomy contracts.  These models deliberately contain references,
not credentials or provider secrets."""
from datetime import datetime
from enum import StrEnum
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field, model_validator
from typing import Generic, TypeVar
from .common import ActorContext, SecretReference

class AgentStatus(StrEnum): draft="draft"; active="active"; paused="paused"; archived="archived"
class ExecutionMode(StrEnum): observe="observe"; recommend="recommend"; act="act"
class TriggerType(StrEnum): manual="manual"; schedule="schedule"
class RunMode(StrEnum): live="live"; dry_run="dry_run"
class ResponseMode(StrEnum): response="response"; actions="actions"; both="both"
class RunStatus(StrEnum): queued="queued"; running="running"; waiting_approval="waiting_approval"; waiting_handoff="waiting_handoff"; succeeded="succeeded"; exhausted="exhausted"; timed_out="timed_out"; cancelled="cancelled"; failed="failed"; suppressed="suppressed"; dead_lettered="dead_lettered"; outcome_unknown="outcome_unknown"
class OverlapPolicy(StrEnum): skip="skip"; buffer_one="buffer_one"

def _reject_secrets(value):
    if isinstance(value, dict):
        for key, item in value.items():
            if any(x in str(key).lower() for x in ("secret", "password", "token", "api_key", "ciphertext")):
                raise ValueError("secret material is not accepted")
            _reject_secrets(item)
    elif isinstance(value, list):
        for item in value: _reject_secrets(item)

class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)
    @model_validator(mode="after")
    def no_secrets(self):
        _reject_secrets(self.model_dump(mode="json")); return self

class Page(BaseModel):
    model_config=ConfigDict(extra="forbid")
    cursor: str|None=None; limit: int=Field(50, ge=1, le=200)

class AgentCreate(StrictModel):
    name: str=Field(min_length=1,max_length=255); handle: str|None=Field(None, pattern=r"^[A-Za-z][A-Za-z0-9_-]{0,31}$"); description: str|None=None; execution_mode: ExecutionMode=ExecutionMode.observe; template_id: UUID|None=None; thread_id: UUID|None=None; concurrency_limit:int=Field(1,ge=1,le=32); queue_limit:int=Field(100,ge=0,le=10000)
class AgentPatch(StrictModel):
    name: str|None=Field(None,min_length=1,max_length=255); handle: str|None=Field(None, pattern=r"^[A-Za-z][A-Za-z0-9_-]{0,31}$"); description:str|None=None; execution_mode:ExecutionMode|None=None; concurrency_limit:int|None=Field(None,ge=1,le=32); queue_limit:int|None=Field(None,ge=0,le=10000)
class AgentResponse(StrictModel):
    id:UUID; thread_id:UUID; thread_title:str|None=None; name:str; handle:str; is_moderator:bool; description:str|None; status:AgentStatus; execution_mode:ExecutionMode; active_version_id:UUID|None; template_id:UUID|None; concurrency_limit:int; queue_limit:int; created_at:datetime; updated_at:datetime

class DraftUpsert(StrictModel):
    optimistic_lock_version:int=Field(ge=1); schema_version:int=Field(1,ge=1); config:dict={}; prompt_template:str=Field("",max_length=100000); tool_selection:list[str]=[]; skill_selection:list[str]=[]; credential_bindings:list[SecretReference]=[]
class DraftResponse(DraftUpsert):
    id:UUID; agent_id:UUID; optimistic_lock_version:int=Field(validation_alias="version",serialization_alias="optimistic_lock_version"); config_hash:str; created_at:datetime; updated_at:datetime
class VersionResponse(StrictModel):
    id:UUID; agent_id:UUID; version:int; schema_version:int; config:dict; prompt_template:str; tool_selection:list; skill_selection:list; credential_bindings:list; config_hash:str; created_at:datetime

class TemplateCreate(StrictModel):
    name:str=Field(min_length=1,max_length=255); description:str|None=None; schema_version:int=Field(1,ge=1); definition:dict={}
class TemplateResponse(StrictModel):
    id:UUID; workspace_id:UUID; name:str; description:str|None; schema_version:int; definition:dict; status:str
class TriggerCreate(StrictModel):
    trigger_type:TriggerType; config:dict={}; is_active:bool=True
    @model_validator(mode="after")
    def valid_schedule(self):
        if self.trigger_type == TriggerType.schedule:
            if not isinstance(self.config.get("cron"),str) or len(self.config["cron"].split()) != 5: raise ValueError("schedule requires a five-field cron")
            from zoneinfo import ZoneInfo
            try: ZoneInfo(self.config.get("timezone","UTC"))
            except Exception as exc: raise ValueError("invalid timezone") from exc
            if self.config.get("overlap","skip") not in {"skip","buffer_one"}: raise ValueError("invalid overlap policy")
        return self
class TriggerResponse(StrictModel):
    id:UUID; workspace_id:UUID; agent_id:UUID; trigger_type:TriggerType; config:dict; is_active:bool
class TriggerPreview(StrictModel):
    timezone:str; occurrences:list[datetime]; assumptions:list[str]
class RunRequest(StrictModel):
    message:str=Field(min_length=1,max_length=100000); mode:RunMode=RunMode.live; trigger_id:UUID|None=None; response_mode:ResponseMode=ResponseMode.both
class RunResponse(StrictModel):
    id:UUID; agent_id:UUID; agent_version_id:UUID; thread_id:UUID; status:RunStatus; mode:RunMode; route:str=""; input_message_id:UUID|None=None; agent_name:str|None=None; agent_handle:str|None=None; trigger_event_id:UUID|None; output_summary:str|None; queued_at:datetime; started_at:datetime|None; completed_at:datetime|None
class EventResponse(StrictModel):
    sequence:int; event_type:str; payload:dict; created_at:datetime
class ForecastResponse(StrictModel):
    frequency_per_day:float; estimated_runs:float; estimated_model_calls:float; estimated_tool_calls:float; assumptions:list[str]
class SchedulePreviewRequest(StrictModel):
    cron:str; timezone:str="UTC"; count:int=Field(5,ge=1,le=20)

T = TypeVar("T")
class AgentPage(BaseModel):
    model_config = ConfigDict(extra="forbid")
    items: list[AgentResponse]
    next_cursor: str | None = None
class RunPage(BaseModel):
    model_config = ConfigDict(extra="forbid")
    items: list[RunResponse]
    next_cursor: str | None = None
class EventPage(BaseModel):
    model_config = ConfigDict(extra="forbid")
    items: list[EventResponse]
    next_cursor: int | None = None
class AuditResponse(StrictModel):
    id: UUID; action: str = Field(validation_alias="event_type", serialization_alias="event_type")
    resource_type: str | None = None; resource_id: str | None = None
    metadata: dict = {}
    created_at: datetime
class AuditPage(BaseModel):
    model_config = ConfigDict(extra="forbid")
    items: list[AuditResponse]
    next_cursor: str | None = None


class HeartbeatDecision(StrEnum): response="response"; action="action"; delegate="delegate"; no_op="no_op"
class HeartbeatOperationalStatus(StrEnum):
    disabled="disabled"; scheduled="scheduled"; evaluating="evaluating"; paused="paused"
    blocked_mode="blocked_mode"; blocked_archived="blocked_archived"; blocked_global="blocked_global"; error="error"


class HeartbeatConfigUpsert(StrictModel):
    enabled: bool
    min_wake_seconds: int = Field(300, ge=30, le=86400)
    max_wake_seconds: int = Field(3600, ge=30, le=604800)
    idle_backoff_factor: float = Field(2.0, ge=1.0, le=10.0)
    expected_revision: int | None = Field(None, ge=1)

    @model_validator(mode="after")
    def _min_le_max(self):
        if self.min_wake_seconds > self.max_wake_seconds:
            raise ValueError("min_wake_seconds must be <= max_wake_seconds")
        return self


class HeartbeatResponse(StrictModel):
    agent_id: UUID
    enabled: bool
    min_wake_seconds: int
    max_wake_seconds: int
    idle_backoff_factor: float
    revision: int
    operational_status: HeartbeatOperationalStatus
    workflow_id: str | None = None
    last_wake_at: datetime | None = None
    last_completed_at: datetime | None = None
    next_wake_at: datetime | None = None
    last_decision: HeartbeatDecision | None = None
    last_run_id: UUID | None = None
    consecutive_noops: int
    last_error: str | None = None
    updated_at: datetime
